### ISSUE TITLE
Tensorflow apply_gradients with horovod

### ISSUE THREAD

**Issue by ankurhanda** [uber/horovod#291]
``` 
trainer = tf.train.AdamOptimizer(learning_rate= self.adam_lr, epsilon=self.adam_epsilon)
self.trainer = hvd.DistributedOptimizer(trainer)
self.adam_op = self.trainer.apply_gradients(self.adam_grads_and_vars)
```

I just want to make sure that this will do parameter averaging. In the horovod code https://github.com/uber/horovod/blob/8a36f3bb6ec089fa69ab033a4eb8ee63ad5ae50e/horovod/tensorflow/__init__.py#L193 I'm not sure if it is doing averaging? I saw another thread here https://github.com/uber/horovod/issues/273. Ideally, I'd like to do apply_gradients on each GPU and then do parameter averaging as that seems natural to me. Can you tell me please if this is happening?

**Comment by ankurhanda** [comment]
Will this do parameter averaging after applying? 
```
self.avg_param = [tf.assign(var, hvd.allreduce(var)) for var in self.var_list]
```

**Comment by alsrgv** [comment]
@ankurhanda, instead of doing parameter averaging, `hvd.DistributedOptimizer()` does gradient averaging:
![image](https://user-images.githubusercontent.com/16640218/41001999-5e22b960-68c7-11e8-975d-198bb5e5086a.png)

Do you see a specific need to do parameter averaging instead?

**Comment by ankurhanda** [comment]
Imagine RL agents training in parallel but each one having their own learning rate so you cannot just do gradient averaging in that case. I added hvd.allreduce(var) that seemed sensible to me but wanted to check with you if you think otherwise? Let me know please.

**Comment by alsrgv** [comment]
I see.  Yes, you can remove `hvd.DistributedOptimizer()` altogether, and after applying the gradients average the variables in the way you described.

**Comment by ankurhanda** [comment]
OK --- cool. That is reassuring. Thank you. I will close the issue.
